src/prefect/experimental/workers/base.py

```diff
@@ -0,0 +1,590 @@
+        raise NotImplementedError(
+            "Workers must implement a method for verifying submitted deployments"
+        )
+
+    @classmethod
+    def __dispatch_key__(cls):
+        if cls.__name__ == "BaseWorker":
+            return None  # The base class is abstract
+        return cls.type
+
+    async def setup(self):
+        """Prepares the worker to run."""
+        self.logger.debug("Setting up worker...")
+        self._runs_task_group = anyio.create_task_group()
+        self._loop_task_group = anyio.create_task_group()
+        self.limiter = (
+            anyio.CapacityLimiter(self.limit) if self.limit is not None else None
+        )
+        self._client = get_client()
+        await self._client.__aenter__()
+        await self._loop_task_group.__aenter__()
+        await self._runs_task_group.__aenter__()
+        self.is_setup = True
+
+    async def teardown(self, *exc_info):
+        """Cleans up resources after the worker is stopped."""
+        self.logger.debug("Tearing down worker...")
+        self.is_setup = False
+        if self._runs_task_group:
+            self._runs_task_group.cancel_scope.cancel()
+            await self._runs_task_group.__aexit__(*exc_info)
+        if self._loop_task_group:
+            self._loop_task_group.cancel_scope.cancel()
+            await self._loop_task_group.__aexit__(*exc_info)
+        if self._client:
+            await self._client.__aexit__(*exc_info)
+        self._runs_task_group = None
+        self._client = None
+
+    async def _on_start(self):
+        """
+        Starts the heartbeat and storage scan loops when the worker starts.
+        """
+        if not self.is_setup or not self._loop_task_group:
+            raise RuntimeError(
+                f"Worker has not been setup. Use `async with {self.__class__.__name__}()...`"
+            )
+        await super()._on_start()
+        # wait for an initial heartbeat to configure the worker
+        await self.heartbeat_worker()
+        # perform initial scan of storage
+        await self.scan_storage_for_deployments()
+        # schedule the heartbeat loop to run every `heartbeat_seconds`
+        self._loop_task_group.start_soon(
+            partial(
+                critical_service_loop,
+                workload=self.heartbeat_worker,
+                interval=self.heartbeat_seconds,
+                printer=self.logger.debug,
+            )
+        )
+        # schedule the storage scan loop to run
+        self._loop_task_group.start_soon(
+            partial(
+                critical_service_loop,
+                workload=self.scan_storage_for_deployments,
+                interval=self.storage_scan_seconds,
+                printer=self.logger.debug,
+            )
+        )
+
+    async def _on_stop(self) -> None:
+        await super()._on_stop()
+
+    def __repr__(self):
+        return f"Worker(pool={self.worker_pool_name!r}, name={self.name!r})"
+
+    async def run_once(self):
```

Comment:
I would consider not using a loop service base class for this. It'd be clearer if this was managed explicitly in your `_loop_task_group`.